File: client/src/services/applicant.js
import axios from '../utils/axios';

export async function createNewApplicant(speciality, experience, salary, resume_text, setError, setShowError, navigate) {
    try {
        const response = await axios.post('/applicants/create', { speciality, experience, salary, resume_text });

        if (response.data) {
            navigate('/applicants')            
            window.location.reload();
            console.log(response.data);
        }
    } catch (error) {
      const errorMessage = error.response.data.detail;
      setError(errorMessage);
      setShowError(true);
      setTimeout(() => {
        setShowError(false);
        setError(null);
      }, 2500);
    }    
}


export async function getPaginatedApplicants(current_page, archived) {
    try {
        const response = await axios.get(`/applicants/?page=${current_page}&archived=${archived}`);

        if (response.data) {
            return response.data;
        }
    } catch (error) {
        console.log(error.response.data.detail);
    }
}


export async function getFilteredApplicants(salary, experience) {
    try {
        const response = await axios.get(`/applicants/filter?salary=${salary}&experience=${experience}`);
        if (response.data) {
            return response.data;
        }
    } catch (error) {
        console.log(error.response.data.detail);
    }
}


export async function searchApplicants(query) {
    try {
      const response = await axios.get(`/applicants/search?query=${query}`);
  
      if (response.data) {
        return response.data;
      }
    } catch (error) {
      console.log(error.response.data.detail);
    }
}  


export async function getApplicantByUserID(user_id) {
    try {
        const response = await axios.get(`/applicants/user/${user_id}`);

        if (response.data) {
            return response.data;
        }
    } catch (error) {
        console.log(error.response.data.detail);
    }
}

export async function deleteApplicant(applicant_id, applicants, setApplicants) {
    try {
        const response = await axios.delete(`/applicants/delete/${applicant_id}`);

        if (response.data) {
            setApplicants(applicants.filter(applicant => applicant.id !== applicant_id));
            console.log(response.data);
        }
    } catch (error) {
        console.log(error.response.data.detail);
    }
}


export async function inArchiveApplicant(applicant_id, applicants, setApplicants) {
    try {
        const response = await axios.patch(`/applicants/in_archive/${applicant_id}`);

        if (response.data) {
            setApplicants(applicants.filter(applicant => applicant.id !== applicant_id));
            console.log(response.data);
        }
    } catch (error) {
        console.log(error.response.data.detail);
    }
}


export async function fromArchiveApplicant(applicant_id, applicants, setApplicants) {
    try {
        const response = await axios.patch(`/applicants/from_archive/${applicant_id}`);

        if (response.data) {
            setApplicants(applicants.filter(applicant => applicant.id !== applicant_id));
            console.log(response.data);
        }
    } catch (error) {
        console.log(error.response.data.detail);
    }
}


export async function inArchiveMySummary(applicant_id, mySummary, setMySummary) {
    try {
        const response = await axios.patch(`/applicants/in_archive/${applicant_id}`);

        if (response.data) {
            setMySummary({ ...mySummary, is_archived: true });
            console.log(response.data);
        }
    } catch (error) {
        console.log(error.response.data.detail);
    }
}


export async function fromArchiveMySummary(applicant_id, mySummary, setMySummary) {
    try {
        const response = await axios.patch(`/applicants/from_archive/${applicant_id}`);

        if (response.data) {
            setMySummary({ ...mySummary, is_archived: false });
            console.log(response.data);
        }
    } catch (error) {
        console.log(error.response.data.detail);
    }
}

File: server/routes/applicant.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from config.database import get_db

from dto.applicant import Applicant as ApplicantDTO
from services.user import get_user_id
from services import applicant as ApplicantService


router = APIRouter()


@router.post("/applicants/create")
async def create_applicant(data: ApplicantDTO, user_id: str = Depends(get_user_id), db: Session = Depends(get_db)):
    return ApplicantService.create_applicant(data, user_id, db)


@router.get("/applicants/user/{user_id}")
async def get_applicant_by_user_id(user_id: str, db: Session = Depends(get_db)):
    return ApplicantService.get_applicant_by_user_id(user_id, db)


@router.get("/applicants")
async def get_paginated_applicants(page: int, archived: bool, user_id: str = Depends(get_user_id), db: Session = Depends(get_db)):
    return ApplicantService.get_paginated_applicants(page, archived, user_id, db)


@router.get("/applicants/filter")
async def get_filtered_applicants(salary: str, experience: str, user_id: str = Depends(get_user_id), db: Session = Depends(get_db)):
    return ApplicantService.get_filtered_applicants(salary, experience, user_id, db)


@router.get("/applicants/search")
async def search_applicants(query: str, user_id: str = Depends(get_user_id), db: Session = Depends(get_db)):
    return ApplicantService.search_applicants(query, user_id, db)


@router.patch("/applicants/in_archive/{applicant_id}")
async def in_archive_applicant(applicant_id: str, user_id: str = Depends(get_user_id), db: Session = Depends(get_db)):
    return ApplicantService.in_archive_applicant(applicant_id, user_id, db)


@router.patch("/applicants/from_archive/{applicant_id}")
async def from_archive_applicant(applicant_id: str, user_id: str = Depends(get_user_id), db: Session = Depends(get_db)):
    return ApplicantService.from_archive_applicant(applicant_id, user_id, db)


@router.delete("/applicants/delete/{applicant_id}")
async def delete_applicant(applicant_id: str, user_id: str = Depends(get_user_id), db: Session = Depends(get_db)):
    return ApplicantService.delete_applicant(applicant_id, user_id, db)
